openactive/openactive-test-suite
# fix(broker): Ditch threads.js for validation multithreading

Fixes https://github.com/openactive/openactive-test-suite/issues/473

What this PR does:

* Removes the threads.js dependency and just uses native worker_threads
* Number of parallel workers is fixed for a given computer - equal to number of cores. Before, this was set to the number of feeds. This should improve performance on a computer which has more cores than feeds and presumably also when a computer has fewer cores (as previously this computer would have been hosting multiple workers per core)
* Rather than enqueueing all items to be validated, items are saved to files (in chunks), which the worker pool then processes.
  * Before, the worst case memory usage might have been holding the entirety of all feeds in memory if validator was slow for some reason (e.g. because the OA context takes 30s to load). This is because it would have been possible to enqueue the entire feed.
  * So, now, items are saved to files and then read one at a time (well, actually: "\<NUMBER OF CORES\> at a time")

## Notes to Reviewer

I recommend starting with:

1. validator-worker-pool.js
2. validator-worker.js
3. validator-inputs.js
4. app.js

## QA

I asked someone who experienced https://github.com/openactive/openactive-test-suite/issues/473 to run this branch, to see if this fixes the issue.

It fixed it!

Since cleaning this up, I have continued to run Broker (`NODE_ENV=dev npm run start-broker`) and see that it works. Here's the results of my latest run:

![Screenshot 2022-02-23 at 17 46 10](https://user-images.githubusercontent.com/2067438/155377076-9c11518f-5b77-482a-932b-ec9f3058261b.png)

![Screenshot 2022-02-23 at 17 48 10](https://user-images.githubusercontent.com/2067438/155377177-394699fe-d74d-4739-8c76-528a441b4ca1.png)
